torch/_functorch/_aot_autograd/subclass_parametrization.py

```diff
@@ -0,0 +1,60 @@
+from typing import List, Tuple
+
+import torch
+from torch.utils._python_dispatch import is_traceable_wrapper_subclass
+
+
+class UnwrapTensorSubclass(torch.nn.Module):
+    def forward(self, *tensors) -> torch.Tensor:  # type: ignore[no-untyped-def]
+        todo: List[torch.Tensor] = list(tensors)
+        for tp, meta, inner_tensors in reversed(self.rebuild_stack):
+            nb_tensor: int = len(inner_tensors)
+            d = {a: b for a, b in zip(inner_tensors, todo[-nb_tensor:])}  # noqa: C416
+            todo = todo[nb_tensor:]
+            rebuilt = tp.__tensor_unflatten__(d, meta, None, None)  # type: ignore[attr-defined]
+            todo.append(rebuilt)
+
+        assert len(todo) == 1
+        return todo[0]
+
+    def right_inverse(self, tensor: torch.Tensor) -> List[torch.Tensor]:
+        assert type(tensor) is not torch.Tensor
+        rebuild_stack = []
+        plain_tensors = []
+        todo = [tensor]
+        while todo:
+            obj = todo.pop()
+            inner_tensors, metadata = obj.__tensor_flatten__()  # type: ignore[attr-defined]
+            rebuild_stack.append((type(obj), metadata, inner_tensors))
+            for attr_name in inner_tensors:
+                val = getattr(obj, attr_name)
+                if type(val) is torch.Tensor:
+                    plain_tensors.append(val)
+                else:
+                    assert isinstance(val, torch.Tensor)
+                    todo.append(val)
+
+        self.rebuild_stack = rebuild_stack
+
+        return plain_tensors
+
+
+def unwrap_tensor_subclass_parameters(model: torch.nn.Module) -> torch.nn.Module:
+    """
+    Model transformation that replaces all the parameters that are subclasses to plain tensors.
+    This reduces runtime overhead of flattening/unflattening the parameters.
+
+    This transformation adds parametrization with `torch.nn.utils.parametrize`.
+    The FQNs of the subclass parameters will be changed and state_dict will become incompatible with the original model.
+    E.g.
+    Original model state_dict: {"p1": torch.testing._internal.TwoTensor}
+    becomes: {"parametrizations.p2.original0": torch.Tensor, "parametrizations.p2.original1": torch.Tensor}
+
+    """
+    name_param: List[Tuple[str, torch.nn.Parameter]] = list(model.named_parameters())
```

Comment:
Why can't this be left an iterator you are only iterating through it?

Reply:
The parameters are changed at every iteration and as a result iteration without list() fails with:
```
    unwrap_tensor_subclass_parameters(m)
  File "/data/users/ivankobzarev/a/pytorch/torch/_functorch/_aot_autograd/subclass_parametrization.py", line 55, in unwrap_tensor_subclass_parameters
    for name, param in name_param:
  File "/data/users/ivankobzarev/a/pytorch/torch/nn/modules/module.py", line 2670, in named_parameters
    yield from gen
  File "/data/users/ivankobzarev/a/pytorch/torch/nn/modules/module.py", line 2607, in _named_members
    for k, v in members:
RuntimeError: dictionary changed size during iteration
```